Moving python allgather_coalesced impl from Py to C

agolynski opened the pull request:
Test Plan:
buck test caffe2/test:c10d
buck test caffe2/test:distributed_gloo

Differential Revision: D18209289

pietern reviewed (changes requested):
In `torch/csrc/distributed/c10d/comm.cpp`, we have implemented `broadcast_coalesced` as a function that takes a process group and a vector of tensors. I think we should do the same for `allgather_coalesced`, since it can be built on top of a regular `allgather`, regardless of the process group implementation. In the current approach, we'd need to duplicate the functionality between the process groups.

mrshenli commented:
> I think we should do the same for allgather_coalesced, since it can be built on top of a regular allgather, regardless of the process group implementation.

@pietern just to make sure we understand your comment. Are you suggesting moving the allgather_coalesced implementation in this PR to  `comm.cpp`? In that case, 1) do we still expose `ProcessGroup::allgather_coalesced` API or does the API only live in `torch.distributed`? 2) I would assume we should do the same for `allreduce_coalesced` too?

pietern commented:
Yes, I think it should live as a separate function. I just checked it out and see there is indeed the `coalesced` variant defined in the `ProcessGroup` base class, but only defined for the Gloo implementation. I think the coalesced variants are generic enough not to need multiple implementations and that they can be built on top of the collectives we expose today. Then both CPU and GPU support and Gloo and NCCL support come for free.

zappa- commented:
> Yes, I think it should live as a separate function. I just checked it out and see there is indeed the `coalesced` variant defined in the `ProcessGroup` base class, but only defined for the Gloo implementation. I think the coalesced variants are generic enough not to need multiple implementations and that they can be built on top of the collectives we expose today. Then both CPU and GPU support and Gloo and NCCL support come for free.

Agreed. As discussed offline, let's do it in a followup PR?

mrshenli commented on torch/lib/c10d/ProcessGroupGloo.cpp:
```diff
@@ -1745,6 +1745,126 @@ std::shared_ptr<ProcessGroup::Work> ProcessGroupGloo::allgather(
+  void allgather_coalesced() {
+    assert(!output_lists.empty());
+    assert(!output_lists[0].empty());
+    assert(!input_list.empty());
```
These checks are already done in `ProcessGroupGloo::allgather_coalesced`, any reason for doing it again here?

zappa- replied:
This is due to linter complaining about possible out of bound access on element 0 of these vectors

mrshenli commented on torch/lib/c10d/Utils.hpp:
```diff
@@ -203,6 +203,20 @@ inline void assertCPU(
+  }
+  const auto& device = tensors[0].device();
+  for (int i = 1; i < tensors.size(); ++i) {
+    if (tensors[i].device() != device) {
```
Were you intended to use `device` var here?

zappa- replied:
Fixed

mrshenli reviewed (approved):
LGTM. I added some cosmetic comments. Based on the offline discussion, we will first land this PR. Then, @zappa- will work on a followup PR to move this implementation to `comm.cpp`.

Thanks a lot for working on this! Please get another stamp from @pietern before landing.

zappa- commented on torch/lib/c10d/Utils.hpp:
```diff
@@ -203,6 +203,20 @@ inline void assertCPU(
+  if (tensors.size() < 2) {
+    return;
+  }
+  const auto& device = tensors[0].device();
```
Fixed

zappa- commented on torch/lib/c10d/Utils.hpp:
```diff
@@ -203,6 +203,20 @@ inline void assertCPU(
+  }
+  const auto& device = tensors[0].device();
+  for (int i = 1; i < tensors.size(); ++i) {
+    if (tensors[i].device() != tensors[0].device()) {
```
Yes, fixed

pietern commented on torch/lib/c10d/ProcessGroupGloo.cpp:
```diff
@@ -1745,6 +1745,126 @@ std::shared_ptr<ProcessGroup::Work> ProcessGroupGloo::allgather(
+
+  assertSameDevice(invalidArgument, input_list);
+
+  // Expect ith tensor of each list from 'output_lists' match ith tensor
```
`i'th` or `tensor at index i`

zappa- replied:
Done

pietern reviewed (approved):
Let's merge this.

@zappa- Can you create issues for the follow up work, one to move this functionality in `comm.cpp`, and one to move `allreduce_coalesced` into `comm.cpp`?

kostmo commented:
## CircleCI build failures summary
As of commit 15b2fc8:
* 0/1 flaky



Here are the [reasons each build failed](https://dr.pytorch.org/commit-details.html?sha1=15b2fc82edcd4ec4e8f56c22501b4e0367f955f9).

---
This comment was automatically generated by [Dr. CI](https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment).
Follow [this link to opt-out](https://dr.pytorch.org/admin/comments-opt-out.html) of these comments for your Pull Requests.


Please report bugs/suggestions on the [GitHub issue tracker](https://github.com/kostmo/circleci-failure-tracker/issues).

zappa- commented:
added issue https://github.com/pytorch/pytorch/issues/29033 to move allgather_coalesced to comm.h. Unfortunately I cannot assign this since I don't have write permissions to repo.

zappa- commented:
Thank you for reviews! Shipping it...

Commit:
Moving python allgather_coalesced impl from Py to C

Summary: Pull Request resolved: https://github.com/pytorch/pytorch/pull/28857

Test Plan:
buck test caffe2/test:c10d
buck test caffe2/test:distributed_gloo

Reviewed By: mrshenli

Differential Revision: D18209289

fbshipit-source-id: f5f394635e60a9d4ce016e366415d90f1daa4423

ezyang commented:
This breaks CUDA builds.

```
Nov 01 18:14:20 /opt/cache/lib/nvcc /var/lib/jenkins/workspace/modules/detectron/spatial_narrow_as_op.cu -c -o /var/lib/jenkins/cpp-build/caffe2/build/modules/detectron/CMakeFiles/caffe2_detectron_ops_gpu.dir//./caffe2_detectron_ops_gpu_generated_spatial_narrow_as_op.cu.o -ccbin /opt/cache/bin/cc -m64 -Dcaffe2_detectron_ops_gpu_EXPORTS -DTH_BLAS_MKL -DONNX_ML=1 -DONNX_NAMESPACE=onnx_torch -DMAGMA_V2 -DIDEEP_USE_MKL -DHAVE_MMAP=1 -D_FILE_OFFSET_BITS=64 -DHAVE_SHM_OPEN=1 -DHAVE_SHM_UNLINK=1 -DHAVE_MALLOC_USABLE_SIZE=1 -D_THP_CORE -DAT_PARALLEL_OPENMP=1 -Xcompiler ,\"-Wno-deprecated\",\"-fvisibility-inlines-hidden\",\"-fopenmp\",\"-DUSE_FBGEMM\",\"-DUSE_QNNPACK\",\"-DUSE_PYTORCH_QNNPACK\",\"-O2\",\"-fPIC\",\"-Wno-narrowing\",\"-Wall\",\"-Wextra\",\"-Wno-missing-field-initializers\",\"-Wno-type-limits\",\"-Wno-array-bounds\",\"-Wno-unknown-pragmas\",\"-Wno-sign-compare\",\"-Wno-unused-parameter\",\"-Wno-unused-variable\",\"-Wno-unused-function\",\"-Wno-unused-result\",\"-Wno-strict-overflow\",\"-Wno-strict-aliasing\",\"-Wno-error=deprecated-declarations\",\"-Wno-error=pedantic\",\"-Wno-error=redundant-decls\",\"-Wno-error=old-style-cast\",\"-fdiagnostics-color=always\",\"-Wno-unused-but-set-variable\",\"-Wno-maybe-uninitialized\",\"-fno-math-errno\",\"-fno-trapping-math\",\"-fPIC\",\"-g\",\"-fno-omit-frame-pointer\",\"-O0\" -DONNX_NAMESPACE=onnx_torch -gencode arch=compute_52,code=sm_52 -Xcudafe --diag_suppress=cc_clobber_ignored -Xcudafe --diag_suppress=integer_sign_change -Xcudafe --diag_suppress=useless_using_declaration -Xcudafe --diag_suppress=set_but_not_used -std=c++11 -Xcompiler -fPIC --expt-relaxed-constexpr --expt-extended-lambda -Wno-deprecated-gpu-targets --expt-extended-lambda -Xfatbin -compress-all -gencode arch=compute_52,code=sm_52 -Xcompiler -fPIC -DCUDA_HAS_FP16=1 -D__CUDA_NO_HALF_OPERATORS__ -D__CUDA_NO_HALF_CONVERSIONS__ -D__CUDA_NO_HALF2_OPERATORS__ -DNVCC -I/usr/local/cuda/include -I/var/lib/jenkins/cpp-build/caffe2/build/aten/src -I/var/lib/jenkins/workspace/aten/src -I/var/lib/jenkins/cpp-build/caffe2/build -I/var/lib/jenkins/workspace -I/var/lib/jenkins/cpp-build/caffe2/build/third_party/gloo -I/var/lib/jenkins/workspace/cmake/../third_party/gloo -I/var/lib/jenkins/workspace/cmake/../third_party/googletest/googlemock/include -I/var/lib/jenkins/workspace/cmake/../third_party/googletest/googletest/include -I/var/lib/jenkins/workspace/third_party/protobuf/src -I/opt/conda/include -I/var/lib/jenkins/workspace/third_party/gemmlowp -I/var/lib/jenkins/workspace/third_party/neon2sse -I/var/lib/jenkins/workspace/cmake/../third_party/benchmark/include -I/var/lib/jenkins/workspace/third_party -I/usr/include -I/var/lib/jenkins/workspace/cmake/../third_party/eigen -I/var/lib/jenkins/workspace/cmake/../third_party/pybind11/include -I/usr/lib/openmpi/include/openmpi/opal/mca/event/libevent2021/libevent -I/usr/lib/openmpi/include/openmpi/opal/mca/event/libevent2021/libevent/include -I/usr/lib/openmpi/include -I/usr/lib/openmpi/include/openmpi -I/opt/rocm/hip/include -I/include -I/var/lib/jenkins/workspace/cmake/../third_party/cub -I/var/lib/jenkins/cpp-build/caffe2/build/caffe2/contrib/aten -I/var/lib/jenkins/workspace/third_party/onnx -I/var/lib/jenkins/cpp-build/caffe2/build/third_party/onnx -I/var/lib/jenkins/workspace/third_party/foxi -I/var/lib/jenkins/cpp-build/caffe2/build/third_party/foxi -I/var/lib/jenkins/workspace/third_party/ideep/mkl-dnn/include -I/var/lib/jenkins/workspace/third_party/ideep/include -I/var/lib/jenkins/workspace/caffe2/../torch/csrc/api -I/var/lib/jenkins/workspace/caffe2/../torch/csrc/api/include -I/var/lib/jenkins/workspace/c10/../ -I/var/lib/jenkins/cpp-build/caffe2/build/third_party/ideep/mkl-dnn/include -I/var/lib/jenkins/workspace/third_party/ideep/mkl-dnn/src/../include -I/var/lib/jenkins/workspace/c10/cuda/../..
Nov 01 18:14:21 In file included from /usr/include/x86_64-linux-gnu/c++/5/bits/c++allocator.h:33:0,
Nov 01 18:14:21                  from /usr/include/c++/5/bits/allocator.h:46,
Nov 01 18:14:21                  from /usr/include/c++/5/string:41,
Nov 01 18:14:21                  from /usr/include/c++/5/stdexcept:39,
Nov 01 18:14:21                  from /usr/include/c++/5/array:38,
Nov 01 18:14:21                  from /usr/include/c++/5/tuple:39,
Nov 01 18:14:21                  from /usr/include/c++/5/mutex:38,
Nov 01 18:14:21                  from /usr/include/c++/5/condition_variable:39,
Nov 01 18:14:21                  from /var/lib/jenkins/workspace/torch/lib/c10d/../c10d/ProcessGroupMPI.hpp:3,
Nov 01 18:14:21                  from /var/lib/jenkins/workspace/torch/lib/c10d/ProcessGroupMPI.cpp:1:

```

ezyang commented:
https://app.circleci.com/jobs/github/pytorch/pytorch/3437322

rohan-varma commented:
I'm getting a similar error to @ezyang: https://circleci.com/api/v1.1/project/github/pytorch/pytorch/3443618/output/106/0?file=true. Is this going to be fixed/reverted?

rohan-varma commented:
Sorry, spoke too soon. it has been reverted by https://github.com/pytorch/pytorch/commit/9041e29d948246c466c27e2c34ed82df62c682d8
